### PR TITLE
Kafka Connect: Commit coordination

### PR DESCRIPTION
**Pull request by bryanck**
This PR is the next stage in submitting the Kafka Connect Iceberg sink connector, and is a follow up to https://github.com/apache/iceberg/pull/8701, https://github.com/apache/iceberg/pull/9466, and https://github.com/apache/iceberg/pull/9641. It includes the commit coordinator and related tests.

Still not included for the sink are the integration tests, distribution build, or docs, which will be added in follow up PRs. For reference, the current sink implementation can be found at https://github.com/tabular-io/iceberg-kafka-connect, and you can read some existing docs at https://github.com/tabular-io/iceberg-kafka-connect/tree/main/docs.

### CODE REVIEW

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java
Comment (ajantha-bhat) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:210]: Since we are only supporting append, do we need this code?

Comment (ajantha-bhat) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:210]: Similar comments for below `RowDelta` block.

File: kafka-connect/kafka-connect/src/test/java/org/apache/iceberg/connect/channel/CoordinatorTest.java
Comment (ajantha-bhat) [inline kafka-connect/kafka-connect/src/test/java/org/apache/iceberg/connect/channel/CoordinatorTest.java:83]: maybe delete file tests can be added when we have delete writers feature added.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (fqaiser94) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:52]: `open` method is called with only the newly assigned partitions. Is there a strong reason to pass just the _newly assigned_ partitions to the `Committer.start` method when the Committer can just retrieve _all_ partitions assigned to this task via `context.assignment` anyway?

I'm also worried we might have a bug here. The Committer implementation uses this `partitions` argument to check if partition 0 of the first topic is assigned to this task and if so, it spawns a Coordinator process. I'm worried that if there was a rebalance where the partition 0 of the first topic doesn't move between tasks, then it would not be included in the `partitions` argument for any Task and thus we could potentially end up with a Connector that doesn't have any Coordinator process running on any Task. Thoughts?

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (fqaiser94) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:81]: Same here, we really don't need to check this on every record?

Feels like we're missing an abstraction here, something like the concept of a `Router` which has a `StaticRouter` implementation and a `DynamicRouter` implementation and only one of those is constructed for the lifetime of a `SinkWriter` based on the `config.dynamicTablesEnabled()` setting.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (fqaiser94) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:121]: IMO we should throw an error instead of dropping the record.
Users can filter out messages easily using an SMT to get the same behaviour, if necessary.
In the future, I imagine we can allow users to supply custom exception-handler implementations which could also allow users to drop records on error.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (fqaiser94) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:124]: Can we support writing to multiple tables if this is a comma separate list of table names? Like we do in "static" mode.

I think this is important because then we can remove regex-matcher configs as it could be implemented entirely as an SMT + dynamic mode.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (fqaiser94) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:112]: nit: is this really "static" routing? I guess it's a static list of tables, but the table/s each message is written to is determined dynamically based on the route value ...

More importantly, is this an important enough use-case to support within the connector? I would strongly prefer if we didn't support this within the connector itself (users can easily implement this by writing an SMT + dynamic mode).

File: kafka-connect/kafka-connect-events/src/test/java/org/apache/iceberg/connect/events/EventTestUtil.java
Comment (bryanck) [reply kafka-connect/kafka-connect-events/src/test/java/org/apache/iceberg/connect/events/EventTestUtil.java:48]: Thanks I changed this.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CatalogUtils.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CatalogUtils.java:49]: I agree, that could be useful, though it may require some discussion to get right, so I'd rather start with this here.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CommitterFactory.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CommitterFactory.java:24]: It isn't needed now, but possibly in the future, to indicate the type of committer to create. The API is public so I designed it with that in mind.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java:83]: We use the connect consumer group now, rather than a separate consumer group that we keep in sync.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:84]: This should never happen, so I changed this to a precondition check.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:210]: I wanted to keep the coordinator capable of handling delete files, so when we do add in delta support, it should be fairly straightforward, and won't require changes to the control message data model.

File: kafka-connect/kafka-connect/src/test/java/org/apache/iceberg/connect/channel/CoordinatorTest.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/test/java/org/apache/iceberg/connect/channel/CoordinatorTest.java:83]: See my comment above, I'd prefer to add this into the coordinator now, rather than later.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:90]: This is here in case future committer implementations want to perform an action on flush. The current committer doesn't do anything when the record collection is null.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:78]: I'm not sure I follow your suggestion. When the KC `close()` or `stop()` lifecycle methods are called, we close the committer and the catalog.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:52]: Thanks, I added a precondition check here to be safer.

Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:52]: That's not my understanding, `open()` will be called with the new assignment, i.e. all assigned topic partitions. See the [javadoc](https://kafka.apache.org/37/javadoc/org/apache/kafka/connect/sink/SinkTask.html#open(java.util.Collection)): "The list of partitions that are now assigned to the task (may include partitions previously assigned to the task)"

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:42]: Thanks for catching this, we can just use the Iceberg version now. I left in the config method in case we want to add to this later.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:118]: We already check this in the config, so I'll just remove this.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:121]: I feel skipping is better than getting into a state where the sink can no longer progress. When we add DLQ support then we could route to that.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:112]: Static in the sense that the list of tables is fixed and doesn't change, rather than deriving it from the record. This feature is in use by some.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CatalogUtils.java
Comment (ajantha-bhat) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CatalogUtils.java:49]: ok. We can move it later.

File: kafka-connect/kafka-connect/src/test/java/org/apache/iceberg/connect/channel/CoordinatorTest.java
Comment (ajantha-bhat) [reply kafka-connect/kafka-connect/src/test/java/org/apache/iceberg/connect/channel/CoordinatorTest.java:83]: Ack

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (ajantha-bhat) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:82]: I think he means just rearrange the code, keep close method just before stop.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (ajantha-bhat) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:121]: +1, we can add a TODO about DLQ here or an issue for tracking it.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (ajantha-bhat) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:81]: Agree. But IMO we should avoid too much internal refactoring which makes extracting code from the Tabular implementation harder. 

We can always refactor once the complete code is in.

**Review by ajantha-bhat — APPROVED** [review]
LGTM. 

The required basic functionality of commit coordinator looks good to me.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:81]: I think this is a good suggestion, and I looked into adding an abstraction here. Ultimately I felt it added complexity, so I agree we should leave it as-is for now.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:82]: I have this order to prevent a running committer from having reference to a closed catalog.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:38]: A while ago there was some discussion about the thread safety of the catalog. At the time we felt we should assume it it thread safe and fix cases where it isn't. I'm not sure if that is still the case, but I'd prefer to assume that, to keep the code simpler for now. I'll raise this in Slack and we can fast follow with an update if needed.

**Comment by bryanck** [comment]
@ajantha-bhat @fqaiser94 Thanks for the reviews! I think I have addressed comments in case you want to take another look. @Fokko @nastra If either of you have time to take a look, that would be very appreciated.

**Comment by ajantha-bhat** [comment]
@rdblue, @danielcweeks, @Fokko, @nastra: Please take a look. It is needed for 1.6.0 milestone.

**Review by fqaiser94 — COMMENTED** [review]
Sorry for the delay, this has been on my radar for a couple of weeks but i don't have a lot of time for reviews currently.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (fqaiser94) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java]: I can tell you that it does happen and there's nothing we can do about it other than throw an error and have the task retry. Precondition check makes sense  👍

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (fqaiser94) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java]: > This is here in case future committer implementations want to perform an action on flush.

What I'm trying to say is that `flush` will never be called since you're overriding the default `preCommit` implementation: https://github.com/apache/kafka/blob/c97d4ce026fdb75f80760bcce00b239db951a481/connect/api/src/main/java/org/apache/kafka/connect/sink/SinkTask.java#L139

So no committer implementation will ever be able to "perform an action on flush" unless you call `flush` yourself somewhere.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (fqaiser94) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:56]: I did read the javadoc and interpreted it differently but I can totally see how you reached your interpretation as well. 

The reason I'm inclined towards my interpretation is because I know `SinkTask.open` is ultimately called by the `ConsumerRebalanceListener.onPartitionsAssigned` method ([here](https://github.com/apache/kafka/blob/958bc0601c7852b8aaf28960280b1f45efd7d850/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L741) in the Kafka source code) and if you look at the javadocs for that method, it says `The list of partitions that are now assigned to the consumer (previously owned partitions will NOT be included, i.e. this list will only include newly added partitions)`. 

I could still be wrong so feel free to test things out or ask in the kafka community.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java
Comment (fqaiser94) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:232]: Alternatively, we could try getting this PR into iceberg-core: https://github.com/apache/iceberg/pull/9860
That would remove the need to do anything special here.

Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:232]: I feel this should be addressed in Iceberg core

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaClientFactory.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaClientFactory.java:59]: I changed this to `putIfAbsent` so it can be overridden.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaClientFactory.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaClientFactory.java:46]: Changed to `putIfAbsent` here also.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Worker.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Worker.java:113]: I removed this, it is no longer necessary.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:90]: This uses daemon threads.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java:106]: Added

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaUtils.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaUtils.java:57]: I added a warn log statement.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java:103]: We need to update the metadata to get the latest generation ID.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java
Comment (danielcweeks) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java:154]: No usages for this, is it necessary?

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java
Comment (danielcweeks) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java:161]: Are we sure this `get()` is safe?  The `allMatch` above will return ture if there are no elements.

**Review by danielcweeks — COMMENTED** [review]
Some minor nits and questions, but overall LGTM.

I think there are still a couple open questions from others (particularly the one on `IcebergSinkTask.open()`, which we probably want to address.

I did see that we renamed a few classes and reduced the visibility in others.  I'm ok with that for now since this isn't really deployable, but we should probably avoid that going forward.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:124]: I think that could be a useful feature, but I don't think it should hold up this PR. If you wanted to open an issue that would be great.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CommitterFactory.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CommitterFactory.java:24]: You're right, I reduced visibility to package private.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java]: I replaced all occurrences of `vtts` .

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java:154]: I removed this, it isn't needed anymore.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java:161]: I changed this to `.orElse(null)`

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:56]: Thanks, it does seem like you're right. I changed this to get the consumer assignment instead.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaUtils.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaUtils.java:66]: I removed the fallback, so it hard fails now.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java
Comment (bryanck) [reply kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java]: I felt it is safer to implement this, given it is a public method, rather than rely on default behavior.

**Comment by bryanck** [comment]
I think I've addressed the latest feedback @fqaiser94 @danielcweeks LMK if there is anything else.

**Review by danielcweeks — APPROVED** [review]

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java
Comment (eshishki) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java:126]: why do actually need to increment? shouldn't we instead decrement to get current offset

**Pull request merged**
